### PR TITLE
Add custom player grounded field for players

### PR DESCRIPTION
**Pull request by Owen1212055**
This method is very nice if you want a simple way to detect if a player is on a block while respecting if the player is on the very edge of the block. 
This adds a custom field to the Player that is only calculated each tick without any interference from packets. This field will only be used in the API.

This new field is accessed via a new Player#isGrounded() method. Better method name suggestions would be appreciated.

### CODE REVIEW

**Comment by electronicboy** [comment]
this doesn't seem to change anything?

**Comment by katekerllenevich** [comment]
I was just about to say the same thing. The `onGround` field in EntityPlayer is still controlled by the client.

**Comment by Owen1212055** [comment]
My goal here was to have the field right after the player entity was ticked. As this field is set when the entity is ticked this would allow be me to get the value before it is modified by anything like packets. I now realize I think I put this in the wrong spot, as the place I put it is updated before the player entity is ticked so this will cause it to be one tick late.

image
I am essentially trying to capture the result of this whilst preventing any modification from packets.

**Comment by katekerllenevich** [comment]
So while that is calculated by the server, it is immediately replaced by what the client says.

```java
PlayerConnection.java

this.player.move(EnumMoveType.PLAYER, new Vec3D(d7, d8, d9));
this.player.setOnGround(packetplayinflying.b()); // CraftBukkit - SPIGOT-5810, SPIGOT-5835: reset by this.player.move
```

**Comment by Owen1212055** [comment]
That is separate however. That sets the field inside of the entity. This adds a special field that isn’t modified by the setter and is only set via the calculation made in the tick method. I misplaced it, it’s supposed to be next to super.tick() in the playerTick method.

**Comment by katekerllenevich** [comment]
In that case should it be a new API method as to not break (possible) backwards compatibility?

**Comment by Owen1212055** [comment]
Perhaps so yes, do you have any idea for what the method would be called? I’m not really sure what to call it to distinguish it enough from its entity counterpart.

**Comment by Owen1212055** [comment]
Okay, I tried making a completely new method. Feedback on the name would be greatly appreciated.

File: Spigot-Server-Patches/0696-Add-custom-grounded-field-for-players.patch [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -540,6 +540,11 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         return server.dispatchCommand(this, command);
+     }
+ 
++    @Override
++    public boolean isGrounded() {
++        return getHandle().playerGrounded;
++    }
++
```
Comment (Owen1212055) [inline Spigot-Server-Patches/0696-Add-custom-grounded-field-for-players.patch]: Forgot paper here woops.

**Comment by lynxplay** [comment]
Do you think there is an merit in also updating the deprecation notice on the `Player#isOnGround` method to point towards this method ?

**Comment by Owen1212055** [comment]
> Do you think there is an merit in also updating the deprecation notice on the `Player#isOnGround` method to point towards this method ?

Wrong commit name but how does it look now?

**Comment by lynxplay** [comment]
Neat! Tho you could simply use `@see Player#isGrounded()`. Shouldn't give each other much.
`@see` also automatically escapes references so no need for a `@link` inside it afaik

**Comment by Owen1212055** [comment]
Updated to 1.17

**Comment by Owen1212055** [comment]
rebased

**Comment by Owen1212055** [comment]
Shoo

**Comment by Owen1212055** [comment]
Thinking about this further, the use of this pull request isn't exactly the best. Doing this urself is simply done by checking to see if the player has any sort of downwards velocity, and then checking if their Y value has changed or not.

**Pull request closed without merging**